src/dbnode/storage/index/fields_terms_iterator.go

```diff
@@ -0,0 +1,200 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+import (
+	"github.com/m3db/m3/src/m3ninx/index/segment"
+	xerrors "github.com/m3db/m3/src/x/errors"
+)
+
+// fieldsAndTermsIterator iterates over all known fields and terms for a segment.
+type fieldsAndTermsIterator interface {
+	// Next returns a bool indicating if there are any more elements.
+	Next() bool
+
+	// Current returns the current element.
+	// NB: the element returned is only valid until the subsequent call to Next().
+	Current() (field, term []byte)
+
+	// Err returns any errors encountered during iteration.
+	Err() error
+
+	// Close releases any resources held by the iterator.
+	Close() error
+
+	// Reset resets the iterator to the start iterating the given segment.
+	Reset(seg segment.Segment, opts fieldsAndTermsIteratorOpts) error
+}
+
+type fieldsAndTermsIteratorOpts struct {
+	iterateTerms bool
+	allowFn      allowFn
+}
+
+func (o fieldsAndTermsIteratorOpts) allow(f []byte) bool {
+	if o.allowFn == nil {
+		return true
+	}
+	return o.allowFn(f)
+}
+
+type allowFn func(field []byte) bool
+
+type fieldsAndTermsIter struct {
+	seg  segment.Segment
+	opts fieldsAndTermsIteratorOpts
+
+	err       error
+	fieldIter segment.FieldsIterator
+	termIter  segment.TermsIterator
+
+	current struct {
+		field []byte
+		term  []byte
+	}
+}
+
+var (
+	fieldsAndTermsIterZeroed fieldsAndTermsIter
+)
+
+var _ fieldsAndTermsIterator = &fieldsAndTermsIter{}
+
+func newFieldsAndTermsIterator(s segment.Segment, opts fieldsAndTermsIteratorOpts) (fieldsAndTermsIterator, error) {
+	iter := &fieldsAndTermsIter{}
+	err := iter.Reset(s, opts)
+	if err != nil {
+		return nil, err
+	}
+	return iter, nil
+}
+
+func (fti *fieldsAndTermsIter) Reset(s segment.Segment, opts fieldsAndTermsIteratorOpts) error {
+	*fti = fieldsAndTermsIterZeroed
+	fti.seg = s
+	fti.opts = opts
+	if s == nil {
+		return nil
+	}
+	fiter, err := s.FieldsIterable().Fields()
+	if err != nil {
+		return err
+	}
+	fti.fieldIter = fiter
+	return nil
+}
+
+func (fti *fieldsAndTermsIter) setNextField() bool {
+	fieldIter := fti.fieldIter
+	if fieldIter == nil {
+		return false
+	}
+
+	for fieldIter.Next() {
+		field := fieldIter.Current()
+		if !fti.opts.allow(field) {
+			continue
+		}
+		fti.current.field = field
+		return true
+	}
+
+	fti.err = fieldIter.Err()
+	return false
+}
+
+func (fti *fieldsAndTermsIter) setNext() bool {
+	// if only need to iterate fields
+	if !fti.opts.iterateTerms {
```

Comment:
nit: I think this might flow a little better if the branch is moved to the `Next()` function instead of calling `setNext` then instantly calling `setNextField`

Reply:
fair, done.